### DISCUSSION TITLE
unsatisfied glibc requirement on SteamOS 3 (Arch based) used by Valve's Steam Deck

### DISCUSSION THREAD

**Discussion by markstinson** [Homebrew/discussions#3489] [Linux] [answered]
tl;dr - Are their any recommendations to address the glibc version challenges given the appliance-like nature of the Steam Deck's rootfs -- SteamOS 3 using v2.33 vs brew's v2.23 ?

linuxbrew is filling a much needed gap on the Steam Deck for folks given its rootfs is read-only and flip-flops between upgrades. (Think of a linux appliance where you can enable rootfs r/w but will result in data loss or worse.)

Some folks are following this well written [Gist for linuxbrew on the Steam Deck](https://gist.github.com/uyjulian/105397c59e95f79f488297bb08c39146) but there's a challenge.

Brew comes with glibc v2.23 where are SteamOS 3 (Arch based) is equipped with glibc v2.33 -- which will change in the future being a rolling linux release.

For example, with brew installed you can `brew install mc` (midnight commander) or `brew install ansible` - no problem.

If you `brew install podman` or one of its deps, `brew install qemu`, the following error appears:
~~~
(deck@steamdeck ~)$ brew install podman
Running `brew update --auto-update`...
glibc: Your system's glibc version is 2.33, and Homebrew's glibc version is 2.23 .
Installing a version of glibc that is older than your system's can break formulae installed from source.
Error: podman: An unsatisfied requirement failed this build.
~~~
Checking https://github.com/Homebrew/homebrew-core/blob/master/Formula/glibc.rb, it's leveraging https://ftp.gnu.org/gnu/glibc/glibc-2.23.tar.gz .

Are their any recommendations to address the glibc version challenges given the appliance-like nature of the Steam Deck's rootfs?

Again, any SteamOS user can `sudo steamos-readonly disable; pacman commands; sudo steamos-readonly enable` but:
* the user will lose installs made on the next OS update so they will need to redo their installs (and the risk that some installs will be broken given /var, /opt, /etc and few other mounts are overlayfs/bindmount to directories in /home/.steamos/offload/ )
* almost makes using brew pointless given the full catalog of pacman installables once you start flipping `steamos-readonly`.

The point of users leveraging brew is to avoid those two huge pain-points. Any assistance/howto dealing the appliance-like SteamOS 3 would be appreciated.

BTW, Valve will be releasing SteamOS 3 for PCs in the very near future. The Deck is just the start. There's community spins already of "SteamOS 3" for those who cannot wait. Brew usage will be growing.

Here's the `brew config` from the Steam Deck's SteamOS 3:
~~~
(deck@steamdeck ~)$ brew config
HOMEBREW_VERSION: 3.5.7
ORIGIN: https://github.com/Homebrew/brew
HEAD: ce8ef89ec0cfe875c48c3a83a843af9074c05930
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: de2ef2def939ddfac571d32dacaf4d83e3cfa81a
Core tap last commit: 35 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8_1/bin/ruby
CPU: octa-core 64-bit zen
Clang: 13.0.0
Git: 2.35.0 => /bin/git
Curl: 7.81.0 => /bin/curl
Kernel: Linux 5.13.0-valve15-1-neptune-02197-gf6ec7ad3762a x86_64 GNU/Linux
OS: SteamOS Holo (Holo)
Host glibc: 2.33
/usr/bin/gcc: 11.1.0
/usr/bin/ruby: N/A
glibc: 2.23
gcc@5: N/A
xorg: N/A

(deck@steamdeck ~)$ brew install gcc
Running `brew update --auto-update`...
Warning: gcc 11.3.0_2 is already installed and up-to-date.
To reinstall 11.3.0_2, run:
  brew reinstall gcc
~~~

**Comment by carlocab** [comment, marked as answer]
Given that your OS already provides glibc 2.33, you shouldn't need to install Homebrew glibc. Any software you `brew install` should work without Homebrew glibc.

I suggest doing `brew uninstall glibc`. You _may_ need to `brew reinstall` things you already have installed after that in case they break.

**Reply by markstinson** [reply]
Thank you. That was the solution.

It was pointed out from the Gist author about "brew on the Steam Deck" to also skip `brew install glibc`.

What I was attempting to do was after their instructions, was to `brew install podman`. The podman installation completed, but now there's a whole other level cruft to wade through.

Brew's podman.rb has `depends_on qemu`. qemu.rb has `depends_on gtk3` (that's there my glibc issue was breaking). Now, the base install of brew on the Steam Deck is about 700MB'ish. After `brew install podman` it was 4.8GB, like OMG ?!? -  (that's not counting the 1GB download cache used as well). That's bigger than some AAA games on the Steam Deck.

Turns out that gtk3 depends_on installs all sorts of X11, wayland, and metric ton of unnecessary GUI rubbish. There is no proper linux distro that requires GUI cruft for CLI qemu and its base utils. Qemu GUI management applications are packaged separately as optionals.

Steam Deck owners wishing to use brew for CLI utils & features will have to evaluate just what exactly they want from brew given its  CLI utils are having GUI dependencies for brew's own X11 and Wayland window managers. Brew may not be an option considering qemu and it's related bins/deps shouldn't take more than a few hundred Megs, and Steam Deck space can be a real premium.

I'm thankful that homebrew exists and has been helpful in the past. I treat it like the 'cygwin for linux' - when you need your user's own Posix install. But it may not be a solution for this CLI use-case which is disappointing.

**Reply by danielnachun** [reply]
@Homebrew/core what do we think about dropping `gtk3` as a dependency for `qemu`?

**Reply by markstinson** [reply]
> @Homebrew/core what do we think about dropping `gtk3` as a dependency for `qemu`?

I think that's a great suggestion. Should it be moved into it's own Issue so the topic isn't lost in my verbose glibc thread here?
